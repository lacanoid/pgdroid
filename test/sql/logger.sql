create table foo (
    id serial primary key,
    label text,
    ctime timestamptz default now()
);

create trigger "$changed" after delete or update on foo 
   for each row execute function dbms.trigger_dml('dbms.dml_log');

insert into foo (label,ctime) values ('Hello world!','2001-01-01 00:00:00');
update foo set label = 'New value';
update foo set label = 'Even more new value';

select relid,usename,pkey,old_data,new_data 
  from dbms.dml_log order by ctime;

select dbms.json_insert('foo', '{"id":-1,"label":"Deuxieme labelle"}', true);
select dbms.json_insert('foo', '{"id":-2,"label":"Troixeme labelle","sublabel":"Ca ira ..."}', true);

select dbms.json_save('foo', '{"id":-3,"label":"Number 4","sublabel":"...."}', true);
select dbms.json_save('foo', '{"id":-3,"label":"Number 5","sublabel":"12345"}', true);

select dbms.json_save2('foo', '{"id":-3,"label":"Number 6","sublabel":"......"}', true);
select dbms.json_save2('foo', '{"id":-3,"label":"Number 7","sublabel":"1234567"}', true);

select dbms.json_insert('foo', '{"label":"Krneki","type":"improved"}',true);
select dbms.json_save2('foo', '{"id":2, "label":"Xrneki","type":"improved+","sublabel":"that"}',true);
#select dbms.json_save2('foo', '{"id":2, "label":"Zrneki","type":"+improved+","sublabel":"that","types":["obi","wan","kenobi"]}',true); --broken

select relid,usename,pkey,old_data,new_data from dbms.dml_log;

