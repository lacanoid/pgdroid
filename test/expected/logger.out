create table foo (
    id serial primary key,
    label text unique,
    ctime timestamptz default now(),
    objid int
);
create unique index foo_objid_idx on foo (objid);
create trigger "$changed" after delete or update on foo 
   for each row execute function dbms.trigger_dml('dbms.log_dml');
insert into foo (label,ctime) values ('Hello world!','2001-01-01 00:00:00');
update foo set label = 'New value';
update foo set label = 'Even more new value';
select relid,usename,pkey,old_data,new_data 
  from dbms.log_dml order by ctime;
 relid | usename |    pkey     |         old_data          |             new_data             
-------+---------+-------------+---------------------------+----------------------------------
 foo   | root    | {"id": "1"} | {"label": "Hello world!"} | {"label": "New value"}
 foo   | root    | {"id": "1"} | {"label": "New value"}    | {"label": "Even more new value"}
(2 rows)

select dbms.json_save('foo', '{"id":-1,"label":"Deuxieme labelle"}', true, true);
NOTICE:  UPDATE foo SET (id, label) = ROW('-1', 'Deuxieme labelle') WHERE id='-1'
NOTICE:  INSERT INTO foo (id, label) VALUES ('-1', 'Deuxieme labelle')
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"id":-2,"label":"Troixeme labelle","sublabel":"Ca ira ..."}', true, true);
NOTICE:  ALTER TABLE foo ADD sublabel text
NOTICE:  UPDATE foo SET (id, label, sublabel) = ROW('-2', 'Troixeme labelle', 'Ca ira ...') WHERE id='-2'
NOTICE:  INSERT INTO foo (id, label, sublabel) VALUES ('-2', 'Troixeme labelle', 'Ca ira ...')
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"id":-3,"label":"Number 4","sublabel":"...."}', true);
NOTICE:  UPDATE foo SET (id, label, sublabel) = ROW('-3', 'Number 4', '....') WHERE id='-3'
NOTICE:  INSERT INTO foo (id, label, sublabel) VALUES ('-3', 'Number 4', '....')
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"id":-3,"label":"Number 5","sublabel":"12345"}', true);
NOTICE:  UPDATE foo SET (id, label, sublabel) = ROW('-3', 'Number 5', '12345') WHERE id='-3'
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"id":-3,"label":"Number 6","sublabel":"......"}', false, true);
NOTICE:  UPDATE foo SET (id, label, sublabel) = ROW('-3', 'Number 6', '......') WHERE id='-3'
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"id":-3,"label":"Number 7","sublabel":"1234567"}', false, true);
NOTICE:  UPDATE foo SET (id, label, sublabel) = ROW('-3', 'Number 7', '1234567') WHERE id='-3'
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"label":"Krneki","type":"improved"}',true,true);
NOTICE:  ALTER TABLE foo ADD type text
NOTICE:  UPDATE foo SET (label, type) = ROW('Krneki', 'improved') WHERE label='Krneki'
NOTICE:  INSERT INTO foo (label, type) VALUES ('Krneki', 'improved')
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"id":2, "label":"Xrneki","type":"improved+","sublabel":"that"}',false,true);
NOTICE:  UPDATE foo SET (id, label, sublabel, type) = ROW('2', 'Xrneki', 'that', 'improved+') WHERE id='2'
 json_save 
-----------
         1
(1 row)

-- select dbms.json_save('foo', '{"id":2, "label":"Zrneki","type":"+improved+","sublabel":"that","types":["obi","wan","kenobi"]}',true); --broken
select dbms.json_save('foo', '{"label":"json_save3 test"}',true);
NOTICE:  UPDATE foo SET (label) = ROW('json_save3 test') WHERE label='json_save3 test'
NOTICE:  INSERT INTO foo (label) VALUES ('json_save3 test')
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"id":3,"label":"json_save3 test ++"}',true);
NOTICE:  UPDATE foo SET (id, label) = ROW('3', 'json_save3 test ++') WHERE id='3'
 json_save 
-----------
         1
(1 row)

select dbms.json_save('foo', '{"id":3,"label":"json_save3 test +++","subtype":"extra"}',true,true);
NOTICE:  ALTER TABLE foo ADD subtype text
NOTICE:  UPDATE foo SET (id, label, subtype) = ROW('3', 'json_save3 test +++', 'extra') WHERE id='3'
 json_save 
-----------
         1
(1 row)

select id,label from foo;
 id |        label        
----+---------------------
  1 | Even more new value
 -1 | Deuxieme labelle
 -2 | Troixeme labelle
 -3 | Number 7
  2 | Xrneki
  3 | json_save3 test +++
(6 rows)

select relid,usename,pkey,old_data,new_data from dbms.log_dml;
 relid | usename |     pkey     |                         old_data                          |                           new_data                           
-------+---------+--------------+-----------------------------------------------------------+--------------------------------------------------------------
 foo   | root    | {"id": "1"}  | {"label": "Hello world!"}                                 | {"label": "New value"}
 foo   | root    | {"id": "1"}  | {"label": "New value"}                                    | {"label": "Even more new value"}
 foo   | root    | {"id": "-3"} | {"label": "Number 4", "sublabel": "...."}                 | {"label": "Number 5", "sublabel": "12345"}
 foo   | root    | {"id": "-3"} | {"label": "Number 5", "sublabel": "12345"}                | {"label": "Number 6", "sublabel": "......"}
 foo   | root    | {"id": "-3"} | {"label": "Number 6", "sublabel": "......"}               | {"label": "Number 7", "sublabel": "1234567"}
 foo   | root    | {"id": "2"}  | {"type": "improved", "label": "Krneki", "sublabel": null} | {"type": "improved+", "label": "Xrneki", "sublabel": "that"}
 foo   | root    | {"id": "3"}  | {"label": "json_save3 test"}                              | {"label": "json_save3 test ++"}
 foo   | root    | {"id": "3"}  | {"label": "json_save3 test ++", "subtype": null}          | {"label": "json_save3 test +++", "subtype": "extra"}
(8 rows)

